src/ffi/mmap.rs

```diff
@@ -0,0 +1,154 @@
+
+use super::{ArrowArray, InternalArrowArray};
+
+#[allow(dead_code)]
+struct PrivateData<T> {
+    // the owner of the pointers' regions
+    data: T,
+    buffers_ptr: Box<[*const std::os::raw::c_void]>,
+    //children_ptr: Box<[*mut ArrowArray]>,
+    dictionary_ptr: Option<*mut ArrowArray>,
+}
+
+fn get_buffer(buffers: &mut VecDeque<IpcBuffer>) -> Result<(usize, usize), Error> {
+    let buffer = buffers
+        .pop_front()
+        .ok_or_else(|| Error::from(OutOfSpecKind::ExpectedBuffer))?;
+
+    let offset: usize = buffer
+        .offset()
+        .try_into()
+        .map_err(|_| Error::from(OutOfSpecKind::NegativeFooterLength))?;
+
+    let length: usize = buffer
+        .length()
+        .try_into()
+        .map_err(|_| Error::from(OutOfSpecKind::NegativeFooterLength))?;
+
+    Ok((offset, length))
+}
+
+// callback used to drop [ArrowArray] when it is exported
+unsafe extern "C" fn release<T>(array: *mut ArrowArray) {
+    if array.is_null() {
+        return;
+    }
+    let array = &mut *array;
+
+    // take ownership of `private_data`, therefore dropping it
+    let private = Box::from_raw(array.private_data as *mut PrivateData<T>);
+    /*for child in private.children_ptr.iter() {
+        let _ = Box::from_raw(*child);
+    }*/
+
+    if let Some(ptr) = private.dictionary_ptr {
+        let _ = Box::from_raw(ptr);
+    }
+
+    array.release = None;
+}
+
+fn mmap_boolean<T: Clone + AsRef<[u8]>>(
+    data: T,
+    node: &Node,
+    block_offset: usize,
+    buffers: &mut VecDeque<IpcBuffer>,
+) -> Result<ArrowArray, Error> {
+    let num_rows: usize = node
+        .length()
+        .try_into()
+        .map_err(|_| Error::from(OutOfSpecKind::NegativeFooterLength))?;
+
+    let null_count: usize = node
+        .null_count()
+        .try_into()
+        .map_err(|_| Error::from(OutOfSpecKind::NegativeFooterLength))?;
+
+    let data_ref = data.as_ref();
+
+    let validity = get_buffer(buffers)?;
+    let (offset, length) = validity;
+
+    let validity = if null_count > 0 {
+        // verify that they are in-bounds and get its pointer
+        Some(data_ref[block_offset + offset..block_offset + offset + length].as_ptr())
+    } else {
+        None
+    };
+
+    let values = get_buffer(buffers)?;
+    let (offset, length) = values;
+
+    // verify that they are in-bounds and get its pointer
+    let values = data_ref[block_offset + offset..block_offset + offset + length].as_ptr();
+
+    // NOTE: this is valid for Boolean, but for others (e.g. Utf8), we need to validate other invariants
+    // or mark this as unsafe
+
+    let buffers_ptr = [validity, Some(values)]
+        .iter()
+        .map(|maybe_buffer| match maybe_buffer {
+            Some(b) => *b as *const std::os::raw::c_void,
+            None => std::ptr::null(),
+        })
+        .collect::<Box<[_]>>();
+    let n_buffers = buffers.len() as i64;
+
+    let mut private_data = Box::new(PrivateData::<T> {
+        data: data.clone(),
```

Comment:
What do we clone here? E.g.` what is `T`?

Reply:
`T` here is anything that implements `Clone + AsRef<[u8]>`, so, for example for `Mmap`, something like `Arc<Mmap<File>>`.

Reply:
Right, I was confused as `Vec<u8>` also implements those. ^^